File: site/scripts/extract-lessons.js
/**
 * Extrai o objeto 'lessons' do caderno-v2.html e converte para JSON.
 * Uso: node extract-lessons.js
 */

const fs = require('fs');
const path = require('path');
const vm = require('vm');

function extractLessons() {
    const siteDir = path.dirname(__dirname);
    const htmlPath = path.join(siteDir, 'caderno-v2.html');
    const outputDir = path.join(siteDir, 'i18n', 'caderno');

    // Criar diretório de saída
    fs.mkdirSync(outputDir, { recursive: true });

    console.log(`[INICIO] Extraindo de ${htmlPath}`);

    // Ler HTML
    const html = fs.readFileSync(htmlPath, 'utf-8');

    // Encontrar o objeto lessons
    const startMarker = 'const lessons = {';
    const startIdx = html.indexOf(startMarker);
    if (startIdx === -1) {
        throw new Error("Objeto 'lessons' não encontrado");
    }

    // Encontrar o final do objeto (contando chaves)
    let braceCount = 0;
    let inString = false;
    let inTemplate = false;
    let stringChar = '';
    let endIdx = startIdx + startMarker.length - 1;

    for (let i = startIdx + startMarker.length - 1; i < html.length; i++) {
        const char = html[i];
        const prevChar = html[i - 1];

        // Detectar strings
        if (!inTemplate && (char === '"' || char === "'") && prevChar !== '\\') {
            if (!inString) {
                inString = true;
                stringChar = char;
            } else if (char === stringChar) {
                inString = false;
            }
            continue;
        }

        // Detectar template literals
        if (char === '`' && prevChar !== '\\') {
            inTemplate = !inTemplate;
            continue;
        }

        // Contar chaves (fora de strings)
        if (!inString && !inTemplate) {
            if (char === '{') braceCount++;
            if (char === '}') {
                braceCount--;
                if (braceCount === 0) {
                    endIdx = i + 1;
                    break;
                }
            }
        }
    }

    const objectCode = html.substring(startIdx + startMarker.length - 1, endIdx);

    // Salvar para debug
    const debugPath = path.join(outputDir, 'debug.js');
    fs.writeFileSync(debugPath, `const lessons = ${objectCode};\nmodule.exports = lessons;`, 'utf-8');

    // Avaliar o código JavaScript usando Function
    let lessons;
    try {
        // Usar Function para criar um escopo limpo
        const evalFunc = new Function(`return ${objectCode}`);
        lessons = evalFunc();
    } catch (e) {
        console.error('[ERRO] Falha ao avaliar JavaScript:', e.message);
        console.log('[INFO] Debug salvo em', debugPath);
        throw e;
    }
    console.log(`[OK] ${Object.keys(lessons).length} lições extraídas`);

    // Criar estrutura de saída
    const output = {
        meta: {
            lang: "pt-BR",
            title: "Formação em IA Aplicada ao Direito",
            subtitle: "Caderno de Exercícios",
            author: "George Marmelstein",
            version: "1.0.0"
        },
        ui: {
            sidebar_title: "IA Aplicada ao Direito",
            sidebar_subtitle: "Caderno de Exercícios",
            theme_light: "Modo claro",
            theme_dark: "Modo escuro",
            prev_lesson: "Lição anterior",
            next_lesson: "Próxima lição",
            exercise: "Exercício",
            tip: "Dica Prática",
            warning: "Armadilha Comum",
            checklist: "Checklist de Verificação",
            copy: "Copiar",
            copied: "Copiado!",
            progress: "Progresso",
            search_placeholder: "Buscar lições...",
            no_results: "Nenhum resultado"
        },
        modules: {
            "modulo0": "Módulo 0 — Preparação",
            "modulo1": "Módulo 1 — Fundamentos",
            "modulo2": "Módulo 2 — Engenharia de Prompt",
            "modulo3": "Módulo 3 — Técnicas Avançadas"
        },
        lessons: lessons
    };

    // Salvar pt.json
    const ptPath = path.join(outputDir, 'pt.json');
    fs.writeFileSync(ptPath, JSON.stringify(output, null, 2), 'utf-8');
    const fileSize = (fs.statSync(ptPath).size / 1024).toFixed(1);
    console.log(`[OK] pt.json (${fileSize} KB)`);

    // Criar esqueletos para en.json e es.json
    for (const lang of ['en', 'es']) {
        const langPath = path.join(outputDir, `${lang}.json`);
        if (!fs.existsSync(langPath)) {
            const skeleton = {
                meta: {
                    lang: lang === 'en' ? "en-US" : "es-ES",
                    title: lang === 'en' ? "AI Training Applied to Law" : "Formación en IA Aplicada al Derecho",
                    subtitle: lang === 'en' ? "Exercise Notebook" : "Cuaderno de Ejercicios",
                    author: "George Marmelstein",
                    version: "1.0.0",
                    status: "pending_translation"
                },
                ui: {
                    sidebar_title: lang === 'en' ? "AI Applied to Law" : "IA Aplicada al Derecho",
                    sidebar_subtitle: lang === 'en' ? "Exercise Notebook" : "Cuaderno de Ejercicios",
                    theme_light: lang === 'en' ? "Light mode" : "Modo claro",
                    theme_dark: lang === 'en' ? "Dark mode" : "Modo oscuro",
                    prev_lesson: lang === 'en' ? "Previous lesson" : "Lección anterior",
                    next_lesson: lang === 'en' ? "Next lesson" : "Próxima lección",
                    exercise: lang === 'en' ? "Exercise" : "Ejercicio",
                    tip: lang === 'en' ? "Practical Tip" : "Consejo Práctico",
                    warning: lang === 'en' ? "Common Pitfall" : "Error Común",
                    checklist: lang === 'en' ? "Verification Checklist" : "Lista de Verificación",
                    copy: lang === 'en' ? "Copy" : "Copiar",
                    copied: lang === 'en' ? "Copied!" : "¡Copiado!",
                    progress: lang === 'en' ? "Progress" : "Progreso",
                    search_placeholder: lang === 'en' ? "Search lessons..." : "Buscar lecciones...",
                    no_results: lang === 'en' ? "No results" : "Sin resultados"
                },
                modules: {
                    "modulo0": lang === 'en' ? "Module 0 — Preparation" : "Módulo 0 — Preparación",
                    "modulo1": lang === 'en' ? "Module 1 — Fundamentals" : "Módulo 1 — Fundamentos",
                    "modulo2": lang === 'en' ? "Module 2 — Prompt Engineering" : "Módulo 2 — Ingeniería de Prompts",
                    "modulo3": lang === 'en' ? "Module 3 — Advanced Techniques" : "Módulo 3 — Técnicas Avanzadas"
                },
                lessons: {}  // Será preenchido pela tradução
            };
            fs.writeFileSync(langPath, JSON.stringify(skeleton, null, 2), 'utf-8');
            console.log(`[OK] ${lang}.json (esqueleto)`);
        }
    }

    console.log('[FIM] Extração concluída');
}

extractLessons();

File: site/scripts/extract-lessons.py
#!/usr/bin/env python3
"""
Extrai o objeto 'lessons' do caderno-v2.html e converte para JSON.
Uso: python extract-lessons.py
"""

import re
import json
import os

def extract_lessons_from_html(html_path):
    """Extrai o objeto lessons do HTML e retorna como dicionário Python."""

    with open(html_path, 'r', encoding='utf-8') as f:
        content = f.read()

    # Encontrar o início do objeto lessons
    match = re.search(r'const lessons = \{', content)
    if not match:
        raise ValueError("Objeto 'lessons' não encontrado no HTML")

    start_idx = match.start() + len('const lessons = ')

    # Encontrar o final do objeto (contando chaves)
    brace_count = 0
    in_string = False
    escape_next = False
    in_template = False
    end_idx = start_idx

    for i, char in enumerate(content[start_idx:], start_idx):
        if escape_next:
            escape_next = False
            continue

        if char == '\\':
            escape_next = True
            continue

        if char == '`':
            in_template = not in_template
            continue

        if in_template:
            continue

        if char == '"' or char == "'":
            # Simplificação: assumir que strings são bem formadas
            pass

        if char == '{':
            brace_count += 1
        elif char == '}':
            brace_count -= 1
            if brace_count == 0:
                end_idx = i + 1
                break

    js_object = content[start_idx:end_idx]

    return js_object, content

def js_to_json(js_obj):
    """Converte objeto JavaScript para JSON válido."""

    # Substituir template literals por strings
    # Padrão: content: `...` -> content: "..."
    result = js_obj

    # Processar template literals (backticks)
    # Esta é uma abordagem simplificada
    in_template = False
    output = []
    i = 0

    while i < len(result):
        char = result[i]

        if char == '`':
            if not in_template:
                in_template = True
                output.append('"')
            else:
                in_template = False
                output.append('"')
            i += 1
            continue

        if in_template:
            # Escapar caracteres especiais dentro de template literals
            if char == '"':
                output.append('\\"')
            elif char == '\n':
                output.append('\\n')
            elif char == '\r':
                output.append('')
            elif char == '\t':
                output.append('\\t')
            elif char == '\\' and i + 1 < len(result) and result[i + 1] == '`':
                # \` dentro de template -> "
                output.append('`')
                i += 2
                continue
            else:
                output.append(char)
        else:
            output.append(char)

        i += 1

    result = ''.join(output)

    # Remover trailing commas antes de } ou ]
    result = re.sub(r',(\s*[}\]])', r'\1', result)

    # Adicionar aspas em chaves não-quoted (ex: number: -> "number":)
    result = re.sub(r'(\s)([a-zA-Z_][a-zA-Z0-9_]*)(\s*:)', r'\1"\2"\3', result)

    return result

def extract_ui_strings(html_content):
    """Extrai strings de UI do HTML."""
    ui = {
        "meta": {
            "lang": "pt-BR",
            "title": "Formação em IA Aplicada ao Direito",
            "subtitle": "Caderno de Exercícios",
            "author": "George Marmelstein",
            "version": "1.0.0"
        },
        "ui": {
            "sidebar_title": "IA Aplicada ao Direito",
            "sidebar_subtitle": "Caderno de Exercícios",
            "theme_light": "Modo claro",
            "theme_dark": "Modo escuro",
            "prev_lesson": "Lição anterior",
            "next_lesson": "Próxima lição",
            "exercise": "Exercício",
            "tip": "Dica Prática",
            "warning": "Armadilha Comum",
            "checklist": "Checklist de Verificação",
            "copy": "Copiar",
            "copied": "Copiado!",
            "progress": "Progresso",
            "search_placeholder": "Buscar...",
            "no_results": "Nenhum resultado encontrado"
        },
        "modules": {
            "modulo0": "Módulo 0 — Preparação",
            "modulo1": "Módulo 1 — Fundamentos",
            "modulo2": "Módulo 2 — Engenharia de Prompt",
            "modulo3": "Módulo 3 — Técnicas Avançadas"
        }
    }
    return ui

def main():
    # Paths
    script_dir = os.path.dirname(os.path.abspath(__file__))
    site_dir = os.path.dirname(script_dir)
    html_path = os.path.join(site_dir, 'caderno-v2.html')
    output_dir = os.path.join(site_dir, 'i18n', 'caderno')

    # Criar diretório de saída
    os.makedirs(output_dir, exist_ok=True)

    print(f"[INICIO] Extraindo lições de {html_path}")

    try:
        # Extrair objeto lessons
        js_lessons, html_content = extract_lessons_from_html(html_path)
        print(f"[OK] Objeto lessons extraído ({len(js_lessons)} caracteres)")

        # Converter para JSON
        json_str = js_to_json(js_lessons)

        # Validar JSON
        try:
            lessons_dict = json.loads(json_str)
            print(f"[OK] JSON válido ({len(lessons_dict)} lições)")
        except json.JSONDecodeError as e:
            # Salvar para debug
            debug_path = os.path.join(output_dir, 'debug_lessons.txt')
            with open(debug_path, 'w', encoding='utf-8') as f:
                f.write(json_str)
            print(f"[ERRO] JSON inválido: {e}")
            print(f"[INFO] Arquivo de debug salvo em {debug_path}")
            return

        # Extrair UI strings
        ui_strings = extract_ui_strings(html_content)

        # Combinar em estrutura final
        output = {
            **ui_strings,
            "lessons": lessons_dict
        }

        # Salvar pt.json
        pt_path = os.path.join(output_dir, 'pt.json')
        with open(pt_path, 'w', encoding='utf-8') as f:
            json.dump(output, f, ensure_ascii=False, indent=2)

        file_size = os.path.getsize(pt_path) / 1024
        print(f"[OK] {pt_path} ({file_size:.1f} KB)")

        # Criar esqueleto para en.json e es.json
        for lang in ['en', 'es']:
            lang_path = os.path.join(output_dir, f'{lang}.json')
            if not os.path.exists(lang_path):
                skeleton = {
                    "meta": {
                        "lang": "en-US" if lang == 'en' else "es-ES",
                        "title": "AI Training Applied to Law" if lang == 'en' else "Formación en IA Aplicada al Derecho",
                        "subtitle": "Exercise Notebook" if lang == 'en' else "Cuaderno de Ejercicios",
                        "author": "George Marmelstein",
                        "version": "1.0.0",
                        "status": "pending_translation"
                    },
                    "ui": {},
                    "modules": {},
                    "lessons": {}
                }
                with open(lang_path, 'w', encoding='utf-8') as f:
                    json.dump(skeleton, f, ensure_ascii=False, indent=2)
                print(f"[OK] {lang_path} (esqueleto criado)")

        print(f"[FIM] Extração concluída")

    except Exception as e:
        print(f"[ERRO] {e}")
        raise

if __name__ == '__main__':
    main()
